attestation-service/verifier/src/se/ibmse.rs

```diff
@@ -0,0 +1,311 @@
+use pv::uv::ConfigUid;
+use serde::{Deserialize, Serialize};
+use serde_with::{base64::Base64, serde_as};
+use std::{env, fs};
+
+const DEFAULT_SE_HOST_KEY_DOCUMENTS_ROOT: &str = "/run/confidential-containers/ibmse/hkds";
+
+const DEFAULT_SE_CERTIFICATES_ROOT: &str = "/run/confidential-containers/ibmse/certs";
+
+const DEFAULT_SE_CERTIFICATE_ROOT_CA: &str = "/run/confidential-containers/ibmse/certs/ca";
+
+const DEFAULT_SE_CERTIFICATE_REVOCATION_LISTS_ROOT: &str =
+    "/run/confidential-containers/ibmse/crls";
+
+const DEFAULT_SE_IMAGE_HEADER_FILE: &str = "/run/confidential-containers/ibmse/hdr/hdr.bin";
+
+const DEFAULT_SE_MEASUREMENT_ENCR_KEY_PRIVATE: &str =
+    "/run/confidential-containers/ibmse/rsa/encrypt_key.pem";
+
+const DEFAULT_SE_MEASUREMENT_ENCR_KEY_PUBLIC: &str =
+    "/run/confidential-containers/ibmse/rsa/encrypt_key.pub";
+
+macro_rules! env_or_default {
+    ($env:literal, $default:ident) => {
+        match env::var($env) {
+            Ok(env_path) => env_path,
+            Err(_) => $default.into(),
+        }
+    };
+}
+
+fn list_files_in_folder(dir: &str) -> Result<Vec<String>> {
+    let mut file_paths = Vec::new();
+
+    for entry in fs::read_dir(dir)? {
+        let entry = entry?;
+        let path = entry.path();
+
+        if path.is_file() {
+            if let Some(path_str) = path.to_str() {
+                file_paths.push(path_str.to_string());
+            }
+        }
+    }
+
+    Ok(file_paths)
+}
+
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SeAttestationResponse {
+    #[serde_as(as = "Base64")]
+    measurement: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    additional_data: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    user_data: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    cuid: ConfigUid,
+    #[serde_as(as = "Base64")]
+    encr_measurement_key: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    encr_request_nonce: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    image_hdr_tags: BootHdrTags,
+}
+
+#[repr(C)]
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SeAttestationClaims {
+    #[serde_as(as = "Base64")]
+    cuid: ConfigUid,
+    #[serde_as(as = "Base64")]
+    user_data: Vec<u8>,
+    version: u32,
+    #[serde_as(as = "Base64")]
+    image_phkh: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    attestation_phkh: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    tag: [u8; 16],
+}
+
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct SeAttestationRequest {
+    #[serde_as(as = "Base64")]
+    request_blob: Vec<u8>,
+    measurement_size: u32,
+    additional_size: u32,
+    #[serde_as(as = "Base64")]
+    encr_measurement_key: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    encr_request_nonce: Vec<u8>,
+    #[serde_as(as = "Base64")]
+    image_hdr_tags: BootHdrTags,
+}
+
+#[derive(Debug)]
+pub struct RealSeVerifier {
+    rsa_private_key: PKey<Private>,
+    rsa_public_key: PKey<Public>,
```

Comment:
I'd prefer EC keys, but that's up to the projects maintainers.

SE attestation uses EC-keys with spec521r1 curve

Reply:
I'm happy to use ECIES approach, after some searching I don't find a handy openssl example, I think pure rust approach is not good for us. Let me add it in my todo. @steffen-eiden WDYT?

Reply:
Instead of reading in a RSA key as pem-file just read in an ec key as pem-file (and rename the variable)
->
`SE_MEASUREMENT_ENCR_KEY_PRIVATE` should contain an EC key in PEm format and you are using EC keys instead of RSA keys.

the PKey abstracts over the exact key type (RSA, EC, HMAC, ....) see [OpenSSL documenatation (C)](https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_new.html)

Reply:
oh, you don't mean ECIES but just replace the RSA key to EC, OK.